### PR TITLE
NHibernate doco review

### PR DESCRIPTION
**Pull request by SzymonPobiega**

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@Particular/nhibernate-persistence-maintainers please review.

File: nservicebus/nhibernate/index.md [hunk 1]
```diff
@@ -58,9 +59,7 @@ partial:code
 
 NServiceBus then picks up the connection setting from the `app.config` from `connectionStrings` and `appSettings` sections. The convention used for `appSettings` does not support defining settings specific for a single persistence concern. If this level of granularity is required use a code-based approach.
 
-NOTE: When using SQL 2012 change the dialect to `MsSql2012Dialect`.
-
-NOTE: Additional dialects are available in the NHibernate.Dialect namespace, [NHibernate documentation.](http://nhibernate.info/doc/)
+NOTE: When using SQL 2012 change the dialect to `MsSql2012Dialect`. Additional dialects are available in the NHibernate.Dialect namespace, [NHibernate documentation.](http://nhibernate.info/doc/)
```
Comment (MarcinHoppe) [inline nservicebus/nhibernate/index.md]: Why are we mentioning the 2012 edition explicitly?

File: nservicebus/nhibernate/accessing-data_limitations_nhibernate_[6,7).partial.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+## Known limitations
+
+When the [transport transactions](/nservicebus/transports/transactions.md) are set to `TransactionScope` mode, due to the way NServiceBus opens sessions (by passing an existing instance of a database connection), it is currently not possible to use NHibernate's second-level cache. Such behavior of NServiceBus is caused by [still-unresolved bug in NHibernate](https://nhibernate.jira.com/browse/NH-3023).
+
+In other transport transaction modes the session is created using a perameterless `OpenSession()` and an explicit NHibernate transaction is used which makes it possible to take advantage of the second-level cache.
```
Comment (MarcinHoppe) [inline nservicebus/nhibernate/accessing-data_limitations_nhibernate_[6,7).partial.md]: Typo: perameterless -> parameterless.

File: nservicebus/handlers/accessing-data.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+- nservicebus/ravendb
+tags:
+ - NHibernate
+ - RavenDB
+ - Persistence
+ - Data access
+ - Saga
+---
+
+In most cases [handlers](/nservicebus/handlers/) are meant to modify the internal state of an application based on the received message. There are multiple ways to do this
```
Comment (MarcinHoppe) [inline nservicebus/handlers/accessing-data.md]: Missing dot at the end of the second sentence.

File: nservicebus/handlers/accessing-data.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+title: Accessing and modifying data
+summary: How to access business data in sync with message consumption and modifications to NServiceBus-controlled data.
+component: Core
+versions: '[5,)'
+related:
+- nservicebus/nhibernate/accessing-data
+- nservicebus/ravendb
+tags:
+ - NHibernate
+ - RavenDB
```
Comment (MarcinHoppe) [inline nservicebus/handlers/accessing-data.md]: Missing Azure Storage Persistence tag?

File: nservicebus/handlers/accessing-data.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+ * Data access context is also shared with the [Saga](/nservicebus/sagas) that might participate in handling a given message
+
+The limitation of this approach is that it requires NServiceBus support for a given data store. Out of the box supported stores include SQL Server and Oracle (via NServiceBus.NHibernate persistence) and RavenDB. 
+
+partial:api
+
+Consult the documentation of a particular persistence for details.
+
+ * [NHibernate](/nservicebus/nhibernate/accessing-data.md)
+ * [RavenDB](/nservicebus/ravendb/#shared-session)
```
Comment (MarcinHoppe) [inline nservicebus/handlers/accessing-data.md]: Ditto.

**Comment by MarcinHoppe** [comment]
Posted some minor comments. The only thing that I miss is that nservicebus/handlers/accessing-data.md focuses on NHibernate and RavenDB, leaving no information about Azure. Maybe a link to "Understanding Transactionality in Azure" doco page would be good?

Otherwise, LGTM.

File: nservicebus/nhibernate/index.md (hunk 1 above)
Comment (SzymonPobiega) [inline nservicebus/nhibernate/index.md]: We default to 2008 dialect. I'll reword it to "releases newer than 2008" maybe

File: Snippets/Core/Core_6/Handlers/DataAccess.cs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        {
+            #region BusinessData-Native-Managed
+
+            public class IdempotencyEnforcer : IHandleMessages<OrderMessage>
+            {
+                public Task Handle(OrderMessage message, IMessageHandlerContext context)
+                {
+                    var session = context.SynchronizedStorageSession.MyOrmSession();
+                    var order = session.Get(message.OrderId);
+                    if (order.HasProcessed(context.MessageId))
```
Comment (SzymonPobiega) [inline Snippets/Core/Core_6/Handlers/DataAccess.cs]: Well, not sure how to name it. The idea here is that when you want to implement the idempotency yourself, you need to somehow persist information about which message has been processed inside the entity itself.

File: Snippets/Core/Core_6/Handlers/DataAccess.cs [hunk 7]
```diff
@@ -0,0 +1,156 @@
+﻿namespace Core5.Handlers
```
Comment (MarcinHoppe) [inline Snippets/Core/Core_6/Handlers/DataAccess.cs]: Core5 -> Core6

File: Snippets/Core/Core_6/Handlers/DataAccess.cs (hunk 6 above)
Comment (MarcinHoppe) [inline Snippets/Core/Core_6/Handlers/DataAccess.cs]: May something along the lines of:

```
if (ChangesFromMessageHaveAlreadyBeenAppliedTo(context.MessageId, order))
{
    //Subsequent handlers are not invoked because the message has already been processed.
    context.DoNotContinueDispatchingCurrentMessageToHandlers();
}
```

**Comment by SzymonPobiega** [comment]
@MarcinHoppe I fixed the code also. Once merged, I'll ping the Azure folks. I would like somebody from docs maintainers to go though it also before merging. @weralabaj @indualagarsamy ?

**Comment by MarcinHoppe** [comment]
@SzymonPobiega I like the idempotency enforcer much better now. Great job 🎉!

File: nservicebus/nhibernate/accessing-data_limitations_nhibernate_[6,7).partial.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+## Known limitations
+
+When the [transport transactions](/nservicebus/transports/transactions.md) are set to `TransactionScope` mode, due to the way NServiceBus opens sessions (by passing an existing instance of a database connection), it is currently not possible to use NHibernate's second-level cache. Such behavior of NServiceBus is caused by [still-unresolved bug in NHibernate](https://nhibernate.jira.com/browse/NH-3023).
+
+In other transport transaction modes the session is created using a parameterless `OpenSession()` and an explicit NHibernate transaction is used which makes it possible to take advantage of the second-level cache.
```
Comment (indualagarsamy) [inline nservicebus/nhibernate/accessing-data_limitations_nhibernate_[6,7).partial.md]: How about:

In order to utilize the second-level cache feature of NHibernate, use a different transport transaction mode instead of `TransactionScope`. NServiceBus uses a parameterless `OpenSession()` to create the session and an explicit NHibernate transaction is used which makes it possible to take advantage of the second-level cache.

File: nservicebus/nhibernate/saga-concurrency.md [hunk 9]
```diff
@@ -15,7 +16,7 @@ One of the most critical things about persistence of sagas is proper concurrency
...
+This approach has a downside of very poor performance in high-contention scenarios where a single saga is accessed by multiple message-processing threads. These threads read the same saga state and process their messages but only one can even succeed persisting the new state. Others experience concurrency violation error and have to try again. In order to overcome this the NHibernate saga persister uses **additional** pessimistic concurrency control using `UPDLOCK` hint. A lock is created when fetching the saga instance from the database and is held till the end of the transaction blocking other threads that try fetching that particular saga. The advantage of this approach that it minimizes the number of retries that would be caused should only the optimistic concurrency was employed. Multiple saga instances can still be processed in parallel.
```
Comment (indualagarsamy) [inline nservicebus/nhibernate/saga-concurrency.md]: >This approach has a downside of very poor performance in high-contention scenarios where a single saga is accessed by multiple message-processing threads.

This approach has a downside of ~~very~~ poor performance in high-contention scenarios where a single saga is accessed by multiple message-processing threads.


> These threads read the same saga state and process their messages but only one can even succeed persisting the new state. 

These threads read the same saga state and process their messages but only one can ~~even~~ succeed *in* persisting the new state. 

>Others experience concurrency violation error and have to try again.

Other threads processing the same message will experience concurrency violation error and will be retried.

>In order to overcome this the NHibernate saga persister uses **additional** pessimistic concurrency control using `UPDLOCK` hint.

this -> what does this imply? the concurrency exceptions? 

>The advantage of this approach that it minimizes the number of retries that would be caused should only the optimistic concurrency was employed. Multiple saga instances can still be processed in parallel.

Not clear. Can you clarify?

File: nservicebus/nhibernate/saga-concurrency.md [hunk 10]
```diff
@@ -26,8 +27,9 @@ snippet:NHibernateConcurrencyRowVersion
...
 
-NOTE: Marking a property with `RowVersion` **does not** disable the pessimistic locking optimization. To switch to pure optimistic concurrency adjust the locking strategy to `Read`.
+NOTE: Marking a property with `RowVersion` **does not** disable the pessimistic locking optimization. All it does is replacing the default optimistic concurrency validation that depends on values of all columns with one that is based on that single explicit version column. To switch to pure optimistic concurrency adjust the locking strategy to `Read`.
 
+In most cases where the saga data table is only ever accessed by the saga persister it is advisable to use an explicit version because the `UPDATE` SQL statement is much simpler and faster. The downside is that it does not detect concurrency violations if the data is updated by some external party that does not conform to the protocol (does not bump the version field when doing updates). If such an external modification is possible it is better to use the default optimistic concurrency validation strategy.
```
Comment (indualagarsamy) [inline nservicebus/nhibernate/saga-concurrency.md]: >The downside is that it does not detect concurrency violations if the data is updated by some external party that does not conform to the protocol (does not bump the version field when doing updates).

The downside is that it does not detect concurrency violations if the data is updated by some external party that does not conform to the protocol i.e. does not bump the version field when doing updates.

>If such an external modification is possible it is better to use the default optimistic concurrency validation strategy.

If such an external modification is possible, it is better to use the default optimistic concurrency validation strategy.

Can you clarify the scenario where an external party is making modifications?

File: nservicebus/handlers/accessing-data.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
+---
+title: Accessing and modifying data
+summary: How to access business data in sync with message consumption and modifications to NServiceBus-controlled data.
```
Comment (SzymonPobiega) [inline nservicebus/handlers/accessing-data.md]: well, you mean I should add `in a message handler` to the summary?

File: nservicebus/handlers/accessing-data.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+
+## Without using NServiceBus persistence
+
+The simplest way to modify the state of the application is to execute the data access code in the handler without using NServiceBus persistence features. Transport transaction mode has to be taken into account when designing such code.
+
+
+### Transport using native transactions
+
+When the selected transport is configured to use native transactions, either in `ReceiveOnly` or `SendsAtomicWithReceive` mode, the data access code in the handler can be executed multiple times for a single message. This can lead to data corruption if that code is not *idempotent* (which is a fancy word for ensuring side effects of the code are the same, no matter how many times the code is invoked). Here's an example of data access code that is not idempotent:
```
Comment (SzymonPobiega) [inline nservicebus/handlers/accessing-data.md]: I added the link and slightly change the desctiption but I don't like "safely receive the same message" because it blurs the definition.

File: Snippets/Core/Core_5/Handlers/DataAccess.cs [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+    public class DataAccess
+    {
+        public class Simple
+        {
+
+            #region BusinessData-Native-NotIdempotent
+
+            public class NonIdempotentHandler : IHandleMessages<AddOrderLine>
+            {
+                public IMyOrm Orm { get; set; }
```
Comment (SimonCropp) [inline Snippets/Core/Core_5/Handlers/DataAccess.cs]: Constructor injection

File: nservicebus/handlers/accessing-data.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+snippet:BusinessData-Native-NotIdempotent
+
+Should the handler be invoked more than once (e.g. because of a transient problem with the transport while committing the receive transaction) the data will get corrupted (duplicate rows will be inserted). The following code shows how to mitigate the problem:
+
+snippet:BusinessData-Native-Idempotent
+
+The downside of this approach is the fact that the code gets more complex. Sometimes it is hard to get enough information from the incoming message to create correct *idempotent* handling logic. This is especially true for commands.
```
Comment (SzymonPobiega) [inline nservicebus/handlers/accessing-data.md]: Added comparison to events.

File: nservicebus/handlers/accessing-data.md [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+Should the handler be invoked more than once (e.g. because of a transient problem with the transport while committing the receive transaction) the data will get corrupted (duplicate rows will be inserted). The following code shows how to mitigate the problem:
+
+snippet:BusinessData-Native-Idempotent
+
+The downside of this approach is the fact that the code gets more complex. Sometimes it is hard to get enough information from the incoming message to create correct *idempotent* handling logic. This is especially true for commands. 
+
+### Transport using distributed transactions
+
+When the selected transport is configured to use distributed transactions via Distributed Transaction Coordinator (DTC) service, the handler is executed within an ambient transaction scope. If the data store supports enlisting in a distributed transaction (e.g. SQL Server, Oracle), the state changes are guaranteed to be applied successfully exactly once, atomically with the message receive operation.
```
Comment (SzymonPobiega) [inline nservicebus/handlers/accessing-data.md]: Reworded.

File: nservicebus/handlers/accessing-data.md [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+
+## Using NServiceBus persistence
+
+Instead of managing everything themselves, users can delegate the management of data store to NServiceBus persistence. This approach has a number of advantages:
```
Comment (SzymonPobiega) [inline nservicebus/handlers/accessing-data.md]: Clarified: managing connections and transactions.

File: nservicebus/handlers/accessing-data.md [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+ * NServiceBus guarantees best practices are followed when it comes to managing data store's connection
+ * Data access context is automatically shared between all handlers executed for a given message, making it easier to guarantee *idempotency* (no partial successes where one handler managed to commit the changes while other didn't)
+ * Data access context is also shared with the [Saga](/nservicebus/sagas) that might participate in handling a given message
+
+The limitation of this approach is that it requires NServiceBus support for a given data store. Out of the box supported stores include SQL Server and Oracle (via NServiceBus.NHibernate persistence) and RavenDB. 
+
+partial:api
+
+Consult the documentation of a particular persistence for details.
```
Comment (SzymonPobiega) [inline nservicebus/handlers/accessing-data.md]: 👍

File: nservicebus/handlers/accessing-data.md [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+ * [NHibernate](/nservicebus/nhibernate/accessing-data.md)
+ * [RavenDB](/nservicebus/ravendb/#shared-session)
+
+### Transport using native transactions
+
+In this mode the NServiceBus-managed data store context can be committed multiple times for a single message and it is up to the user to guarantee *idempotency*. Unlike in the basic example, though, the data store context is shared between the handlers so there might be one (possibly generic) handler that takes care of the *idempotency*, allowing others to focus on pure business problem.
```
Comment (SzymonPobiega) [inline nservicebus/handlers/accessing-data.md]: reworded.

File: nservicebus/handlers/accessing-data.md [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+snippet:BusinessData-Native-Managed
+
+
+### Outbox
+
+When using the [Outbox](/nservicebus/outbox) NServiceBus itself guarantees the *idempotency* of data access operations executed via `SynchronizedStorageSession`. The Outbox can be seen as a generic implementation of the *idempotency*-ensuring handler from the previous section.
```
Comment (SzymonPobiega) [inline nservicebus/handlers/accessing-data.md]: SynchronizedStorageSession is defined in a partial, earlier in this doco. I switch to partials for outbox because SynchronizedStorageSession  is only available in 6.

File: nservicebus/nhibernate/accessing-data_limitations_nhibernate_[6,7).partial.md (hunk 8 above)
Comment (SzymonPobiega) [inline nservicebus/nhibernate/accessing-data_limitations_nhibernate_[6,7).partial.md]: Well, I don't thing 2nd level cache is a criterium to use when deciding on transport transaction mode. I'd rather state that is works in other modes than have people switch to lower modes and then discover then have duplicate messages.

File: nservicebus/nhibernate/saga-concurrency.md (hunk 9 above)
Comment (SzymonPobiega) [inline nservicebus/nhibernate/saga-concurrency.md]: Reworded.

**Comment by SzymonPobiega** [comment]
@indualagarsamy I've applies your suggestions.

File: nservicebus/handlers/accessing-data.md [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+ - Data access
+ - Saga
+---
+
+In most cases [handlers](/nservicebus/handlers/) are meant to modify the internal state of an application based on the received message. There are multiple ways to do this.
+
+
+## Without using NServiceBus persistence
+
+The simplest way to modify the state of the application is to execute the data access code in the handler without using NServiceBus persistence features. Transport transaction mode has to be taken into account when designing such code.
```
Comment (SimonCropp) [inline nservicebus/handlers/accessing-data.md]: @SzymonPobiega can u address this

File: nservicebus/handlers/accessing-data.md (hunk 11 above)
Comment (SimonCropp) [inline nservicebus/handlers/accessing-data.md]: @indualagarsamy can u answer this question

File: nservicebus/handlers/accessing-data.md (hunk 20 above)
Comment (SzymonPobiega) [inline nservicebus/handlers/accessing-data.md]: @SimonCropp @indualagarsamy I reworded this para a bit
